src/Evolution/Systems/NewtonianEuler/TimeDerivative.hpp

```diff
@@ -0,0 +1,150 @@
+    const tnsr::I<DataVector, Dim>& momentum_density,
+    const Scalar<DataVector>& energy_density,
+    const tnsr::I<DataVector, Dim>& velocity,
+    const Scalar<DataVector>& pressure) noexcept;
+}  // namespace detail
+
+/*!
+ * \brief Compute the time derivative of the conserved variables for the
+ * Newtonian Euler system
+ */
+template <size_t Dim, typename InitialDataType>
+struct TimeDerivative {
+ private:
+  struct EnthalpyDensity : db::SimpleTag {
+    using type = Scalar<DataVector>;
+  };
+
+  using SourceTerm = typename InitialDataType::source_term_type;
+  using argument_tags_flux =
+      tmpl::list<Tags::MomentumDensity<Dim>, Tags::EnergyDensity,
+                 Tags::Velocity<DataVector, Dim>, Tags::Pressure<DataVector>>;
+
+ public:
+  using temporary_tags = tmpl::list<EnthalpyDensity>;
+  using argument_tags = tmpl::append<
+      argument_tags_flux,
+      tmpl::conditional_t<std::is_same_v<SourceTerm, Sources::NoSource>,
+                          tmpl::list<>,
+                          tmpl::push_front<typename SourceTerm::argument_tags,
+                                           Tags::SourceTerm<InitialDataType>>>>;
+
+  template <typename... SourceTermArgs>
+  static void apply(
+      // Time derivatives returned by reference. All the tags in the
+      // variables_tag in the system struct.
+      const gsl::not_null<Scalar<DataVector>*> dt_mass_density,
+      const gsl::not_null<tnsr::I<DataVector, Dim>*> dt_momentum_density,
+      const gsl::not_null<Scalar<DataVector>*> dt_energy_density,
+
+      // Fluxes returned by reference. Listed in the system struct as
+      // flux_variables.
+      const gsl::not_null<tnsr::I<DataVector, Dim>*> mass_density_cons_flux,
+      const gsl::not_null<tnsr::IJ<DataVector, Dim>*> momentum_density_flux,
+      const gsl::not_null<tnsr::I<DataVector, Dim>*> energy_density_flux,
+
+      // Temporaries returned by reference. Listed in temporary_tags above.
+      const gsl::not_null<Scalar<DataVector>*> enthalpy_density,
+
+      // Arguments listed in argument_tags above
+      const tnsr::I<DataVector, Dim>& momentum_density,
+      const Scalar<DataVector>& energy_density,
+      const tnsr::I<DataVector, Dim>& velocity,
+      const Scalar<DataVector>& pressure, const SourceTerm& source,
+      const SourceTermArgs&... source_term_args) noexcept {
+    detail::fluxes_impl(mass_density_cons_flux, momentum_density_flux,
+                        energy_density_flux, enthalpy_density, momentum_density,
+                        energy_density, velocity, pressure);
+    if constexpr (not std::is_same_v<SourceTerm, Sources::NoSource>) {
+      sources_impl(
+          std::make_tuple(dt_mass_density, dt_momentum_density,
+                          dt_energy_density),
+          typename InitialDataType::source_term_type::sourced_variables{},
+          source, source_term_args...);
+    }
+  }
+
+  template <typename... SourceTermArgs>
```

Comment:
Well that's a great question! Clang doesn't need it (haven't tested if GCC does) so I'm removing it. Basically the compiler deduces that `SourceTermArgs` is an empty parameter pack, which is allowed. However, by making the function a template you can invoke ADL and cause delayed parsing and instantiations, which is sometimes a useful trick (though should be avoided because it's too clever to be readable :) )

Reply:
Thanks for clarifying!